### DISCUSSION TITLE
Display tags at the beginning of each section

### DISCUSSION THREAD

**Discussion by qingshuizheng** [ahyatt/ekg#35] [Ideas]
Hi @ahyatt ,

Is it a good idea to display tags at the beginning of each section, instead of at the end?
It seems to me that it will be easier to get what is it all about in the long paragraph.

It would be something like this:

![image](https://user-images.githubusercontent.com/78839107/232566922-85c42251-71cf-4e8a-a12d-f13c2463e70c.png)

Best, 
Zheng

**Comment by ahyatt** [comment]
I'm curious to see what others think.  For me, I prefer at the bottom, but I don't have any real reason for that preference.  I'm thinking about ways to customize this now, which may appear in the next version.

**Comment by cuprum** [comment]
I already wrote about the readability and also realized that I prefer tags at the top of the notes. At least it looks more consistently to me - tags at the top on the capture buffer and tag's buffer.

Here's the code I sketched out:

``` emacs-lisp
(defcustom ekg-note-view '("%text" "%line" "%tags" "%line" "%faced-line" "%line")
  "Parts of note..."
  :type '(set string)
  :group 'ekg)

(defface ekg-faced-line
  '((default :strike-through t))
  "....")

(defun ekg--print-note-parts (note-view note)
  (let ((note-parts `("%text" ,(ekg-displayable-note-text note)
		      "%tags" ,(ekg-tags-display (ekg-note-tags note))
                      "%line" "\n"
                      "%faced-line" #(" " 0 1 (face 'ekg-faced-line display (space :align-to right)))))
	output)
    (setq output (cl-loop for el in note-view
			  collect (lax-plist-get note-parts el)))
    (mapcar #'insert output)))

(defun ekg-display-note (note)
  "Display NOTE in buffer."
  (when (ekg--should-show-id-p (ekg-note-id note))
    (insert
     (propertize
      (format "[%s]\n" (ekg-note-id note))
      'face 'ekg-resource)))
  (ekg--print-note-parts ekg-note-view note))
```
This is just proof of concept. The code requires testing, but the general idea is clear - user can customize the appearance of notes. Notes can look like

![1](https://user-images.githubusercontent.com/532546/232836934-e2b488f2-3d45-4049-9c67-4de88055653e.png)

or

![2](https://user-images.githubusercontent.com/532546/232837038-3d566cd1-dc13-4b56-8ce9-c8cffb64c2a9.png)

**Reply by qingshuizheng** [reply]
@ahyatt 
Usually when ppl read a blog post, most of the time (i assume) they see tags just below the title. Only in very few cases, blogger put them at bottom. I would like to see any customizable options in the coming commits.

@cuprum 
Consistency is one of my considerations too. 
Your solution looks great! I'm gonna try it a moment later.

**Reply by ahyatt** [reply]
This is an interesting way to approach this, thanks for sharing.  I am working on a way to be able to have display functions per "type", so let me also use your idea of an ordered list of types and newlines/hlines that will help us order them to control the rendering.  However, since we don't know of all types, we'll also have to have a way to denote the place to put misc data that you didn't know about, such as if we had data about a "person" type, where in the list would that get rendered?

**Reply by cuprum** [reply]
@ahyatt 
Of course use it, no problem. As for the unknown types, unfortunately, I can't say anything useful here. Maybe some kind of template system? For "person" type a user will specify %person, for "weather" type - %weather and so on. The appearance of the template may be different.
I saw a similar thing in org-roam: https://github.com/org-roam/org-roam/wiki/User-contributed-Tricks#showing-the-number-of-backlinks-for-each-node-in-org-roam-node-find. As far as I understand, under the hood there is used machinery that matches the template and the end in the function name.

**Reply by ahyatt** [reply]
You can try the latest on the `inlines` branch.  I ended up using a templating system, which fit well with the inline system that I created.  I also added documentation, so you can read the info file or `doc/ekg.org` to see the details.

**Reply by qingshuizheng** [reply]
@ahyatt I haven't looked into the code in ekg/inlines branch, just experimented the `ekg-display-note-template`, and it turns out great, thanks for making this happen.

With this line:
`
(setq ekg-display-note-template "%n(tagged)%n(id)%n(titled)%n(text 500)%n(other)")
`

![image](https://user-images.githubusercontent.com/78839107/233858709-3f3d11dd-ac3d-4312-b0c4-fce50ec6316c.png)

**Reply by cuprum** [reply]
@ahyatt I looked it up, everything works as expected, great job!

**Comment by ahyatt** [comment]
Thanks, since this seems reasonably done, I'm going to close this out.  If you see anything wrong with it, please open a bug.  And thanks for starting this conversation!
